[ONNX] Allow exporter to export SDPA to Attention onnx operator

kkhode opened the pull request:
Fixes [#149662](https://github.com/pytorch/pytorch/issues/149662)

kkhode committed:
Allow exporter to export SDPA to Attention onnx operator

justinchuby commented on torch/onnx/_internal/exporter/_torchlib/ops/nn.py:
```diff
@@ -46,3 +47,68 @@ def aten_group_norm(
+
+    if scale is None:
+        scale = nn._attention_scale(query)
+    scale = op18.CastLike(scale, query)
```
Prefer using the same opset within the same function. 

```suggestion
    scale = op23.CastLike(scale, query)
```

kkhode replied:
Added change in latest commit.

justinchuby commented on torch/onnx/_internal/exporter/_torchlib/ops/nn.py:
```diff
@@ -46,3 +47,68 @@ def aten_group_norm(
+    where Q, K, V are the query, key, and value tensors, respectively.
+    L is the target sequence length, S is the source sequence length, and E is the embedding size.
+    """
+    # Use trace_only to handle optional inputs
```
```suggestion
```

trace_only is now the default recommended path so we can remove the comment

kkhode replied:
Added change in latest commit.

justinchuby commented:
Than you! Added a few comments

kkhode committed:
Duplicated functions available in onnxscript.function_libs.torch_lib.ops.nn ...

Duplicated attention related functions can now be called with varied opsets
Added reviewed naming and comment changes

kkhode commented on torch/onnx/_internal/exporter/_torchlib/ops/nn.py:
```diff
@@ -46,3 +47,68 @@ def aten_group_norm(
     )
+
+@onnx_impl(aten.scaled_dot_product_attention.default, trace_only=True, opset_introduced=23)
+def scaled_dot_product_attention(
```
Added change in latest commit.

justinchuby commented on torch/onnx/_internal/exporter/_torchlib/ops/nn.py:
```diff
@@ -7,7 +7,7 @@
 from typing import Optional
 
-import onnxscript.function_libs.torch_lib.ops.nn as nn
+from onnxscript.values import Opset
```
Wrap this import in if TYPE_CHECKING: because it is not used in runtime?

kkhode committed:
Add a unit test for attention onnx op & add type_checking test

kkhode committed:
Decouple graph sanity and accuracy check

kkhode commented on test/onnx/exporter/test_small_models_e2e.py:
```diff
@@ -629,6 +630,20 @@ def forward(self, x):
+        value = torch.rand(32, 8, 128, 64, dtype=torch.float16)
+
+        onnx_program = self.export(Model(), (query, key, value), opset_version=23)
+        self.assertIn("Attention", [node.op_type for node in onnx_program.model.graph])
```
Addressed in latest commit.

kkhode committed:
Fix lint issues

justinchuby commented on torch/onnx/_internal/exporter/_torchlib/ops/nn.py:
```diff
@@ -46,3 +51,196 @@ def aten_group_norm(
+            )
+        Y, _, _, _ = op23.Attention(
+            query, key, value, attn_mask=attn_mask, scale=scale, q_num_heads=query.shape[3], kv_num_heads=key.shape[3],
+            is_causal=(1 if is_causal else 0)
```
```suggestion
            is_causal=is_causal
```

kkhode commented on torch/onnx/_internal/exporter/_torchlib/ops/nn.py:
```diff
@@ -96,7 +97,7 @@ def aten_scaled_dot_product_attention_23(
         Y, _, _, _ = op23.Attention(
             query, key, value, attn_mask=attn_mask, scale=scale, q_num_heads=query.shape[3], kv_num_heads=key.shape[3],
-            is_causal=(1 if is_causal == True else 0)
+            is_causal=(1 if is_causal else 0)
```
It is fine. But schema and type checking recommends int. Didn't want the logic to stumble and fall on type checking or schema checks, if they're added.

kkhode committed:
Update torch/onnx/_internal/exporter/_torchlib/ops/nn.py

Co-authored-by: Justin Chu <justinchuby@users.noreply.github.com>

kkhode committed:
Fix lint issues by following lintrunner to a tee

justinchuby commented:
@pytorchbot merge

justinchuby committed:
Update nn.py

kkhode committed:
Fix incorrect assertion logic for tensor shape

justinchuby commented on torch/onnx/_internal/exporter/_torchlib/ops/nn.py:
```diff
@@ -46,3 +51,223 @@ def aten_group_norm(
+) -> TFloat:
+    # Swap the last two axes of key
+    key_shape = op.Shape(key)
+    key_last_dim = op.Slice(key_shape, [-1], op.Constant(value_ints=[_INT64_MAX]))
+    key_second_last_dim = op.Slice(key_shape, [-2], [-1])
+    key_first_dims = op.Slice(key_shape, op.Constant(value_ints=[_INT64_MIN]), [-2])
```
That's fine

justinchuby committed:
Update nn.py

kkhode committed:
Use shape attributes to slice since opset23 usage is guaranteed

titaiwangms commented on test/onnx/exporter/test_small_models_e2e.py:
```diff
@@ -629,6 +630,35 @@ def forward(self, x):
+        value = torch.rand(32, 8, 128, 64, dtype=torch.float16)
+
+        onnx_program = self.export(Model(), (query, key, value), opset_version=23)
+        onnx_testing.assert_onnx_program(onnx_program, atol=1e-3, rtol=1)
```
I forgot it's a function op. ort should work.

justinchuby replied:
The dilemma is ort in pytorch ci is 1.18 because pytorch ci runs on python 3.9. I think we can conditionally fail this test depending on the onnxruntime version?

titaiwangms replied:
Does 1.18 not work with function op?

justinchuby replied:
It should fail with opset 23 not supported

titaiwangms reviewed (approved):
We will need to revisit once @xadupre done the Attention implementation in ORT.

justinchuby commented:
@pytorchbot merge

justinchuby commented:
@pytorchbot merge

justinchuby commented:
@pytorchbot merge -i

justinchuby commented:
@pytorchbot merge -f "lint pipelines are timing out due to docker errors"

justinchuby commented on torch/onnx/_internal/exporter/_torchlib/ops/nn.py:
```diff
@@ -46,3 +52,221 @@ def aten_group_norm(
+                "SDPA (MHA) requires q_num_heads = kv_num_heads"
+            )
+        Y, _, _, _ = op23.Attention(
+            query,
```
onnx query shape is `(batch_size, q_num_heads, q_sequence_length, head_size)` but it looks like pytorch can have an arbitrary number of dimensions in the middle 
![image](https://github.com/user-attachments/assets/eaea4d8e-4b15-4239-945d-9e5ba1deff00)

Do we need to reshape the inputs when they are not rank 4?
